sklearn/experimental/_column_transformer.py

```diff
@@ -0,0 +1,212 @@
+        self : ColumnTransformer
+            This estimator
+        """
+        try:
+            return super(ColumnTransformer, self).fit(X, y=y)
+        except ValueError as e:
+            if "Got X with X.ndim=1. Reshape your data" in str(e):
+                raise ValueError(_ERR_MSG_1DCOLUMN)
+            else:
+                raise e
+
+    def fit_transform(self, X, y=None, **fit_params):
+        """Fit all transformers, transform the data and concatenate results.
+
+        Parameters
+        ----------
+        X : array-like or DataFrame of shape [n_samples, n_features]
+            Input data, of which specified subsets are used to fit the
+            transformers.
+
+        y : array-like, shape (n_samples, ...), optional
+            Targets for supervised learning.
+
+        Returns
+        -------
+        X_t : array-like or sparse matrix, shape (n_samples, sum_n_components)
+            hstack of results of transformers. sum_n_components is the
+            sum of n_components (output dimension) over transformers.
+        """
+        try:
+            return super(ColumnTransformer, self).fit_transform(X, y=y,
+                                                                **fit_params)
+        except ValueError as e:
+            if "Got X with X.ndim=1. Reshape your data" in str(e):
+                raise ValueError(_ERR_MSG_1DCOLUMN)
+            else:
+                raise e
+
+    def transform(self, X):
+        """Transform X separately by each transformer, concatenate results.
+
+        Parameters
+        ----------
+        X : array-like or DataFrame of shape [n_samples, n_features]
+            Input data, of which specified subsets are used to fit the
+            transformers.
+
+        Returns
+        -------
+        X_t : array-like or sparse matrix, shape (n_samples, sum_n_components)
+            hstack of results of transformers. sum_n_components is the
+            sum of n_components (output dimension) over transformers.
+        """
+        try:
+            return super(ColumnTransformer, self).transform(X)
+        except ValueError as e:
+            if "Got X with X.ndim=1. Reshape your data" in str(e):
+                raise ValueError(_ERR_MSG_1DCOLUMN)
+            else:
+                raise e
+
+
+def _get_column(X, key):
+    """
+    Get feature column(s) from input data X.
+
+    Supported input types (X): numpy arrays, sparse arrays and dataframes
+
+    Supported key types (key):
+    - scalar: output is 1D
+    - lists, slices: output is 2D
+
+    Supported key data types:
+
+    - integer (positional):
+        - supported for (sparse) arrays or dataframes
+    - string (key-based):
+        - only supported for dataframes
+        - So no keys other than strings are allowed (while in principle you
+          can use any hashable object as key).
+
+    """
+    if X is None:
+        return X
+
+    # check whether we have string column names or integers
```

Comment:
I think even this would be worthwhile:
```python
def _check_key_type(key, superclass):
    if isinstance(key, superclass):
        return True
    if isinstance(key, slice):
        return isinstance(key.start, (superclass, None)) and isinstance(key.stop, (superclass, None))
    if isinstance(key, list):
        return all(isinstance(x, superclass) for x in key)
    if hasattr(key, 'dtype'):
        return key.dtype.kind == 'i'
    return False
```